src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -114,19 +142,21 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
             // increment the in-flight of that solution until we decrement it at the end of our try/finally block.
             var (inFlightSolution, solutionTask) = await AcquireSolutionAndIncrementInFlightCountAsync().ConfigureAwait(false);
 
+            Exception? exception = null;
             try
             {
-                return await ProcessSolutionAsync(inFlightSolution, solutionTask).ConfigureAwait(false);
+                var solution = await GetSolutionAsync(inFlightSolution, solutionTask).ConfigureAwait(false);
+                return new PinnedSolution(this, inFlightSolution, solution);
             }
-            catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, cancellationToken, ErrorSeverity.Critical))
+            catch (Exception ex) when ((exception = ex) == null)
             {
-                // Any non-cancellation exception is bad and needs to be reported.  We will still ensure that we cleanup
-                // below though no matter what happens so that other calls to OOP can properly work.
                 throw ExceptionUtilities.Unreachable();
             }
             finally
             {
-                await DecrementInFlightCountAsync(inFlightSolution).ConfigureAwait(false);
+                // Ensure we cleanup if an exception occurred.  Otherwise, we'll never release this data.
+                if (exception != null)
+                    await DecrementInFlightCountAsync(inFlightSolution).ConfigureAwait(false);
```

Comment:
If the intent is "only run this if we didn't throw an exception", them just move it out of the finally block? I'm confused what's going on here.

Reply:
so i don't disagree.  but my understanding from @tmat is that this is the right pattern to ensure proper async callstacks.  if we move this into the `catch` then we get back stacks (even if we use `throw;`).  @tmat can you elaborate here?